diff -urNp -x '*.orig' cvs-1.12.13.org/configure.in cvs-1.12.13/configure.in
--- cvs-1.12.13.org/configure.in	2005-10-03 15:34:45.000000000 +0200
+++ cvs-1.12.13/configure.in	2023-02-02 09:00:39.745239468 +0100
@@ -514,13 +514,12 @@ dnl for debugging code
 CVS_FUNC_PRINTF_PTR
 
 # Try to find connect and gethostbyname.
-AC_CHECK_LIB(nsl, main)
 AC_SEARCH_LIBS(connect, xnet socket inet,
   AC_DEFINE(HAVE_CONNECT, 1,
 [Define if you have the connect function.]))
 dnl no need to search nsl for gethostbyname here since we should have
 dnl just added libnsl above if we found it.
-AC_SEARCH_LIBS(gethostbyname, netinet)
+AC_SEARCH_LIBS(gethostbyname, netinet nsl)
 
 AC_SUBST(cvs_client_objects)
 
