diff -urN cvs-1.11.11.org/src/client.c cvs-1.11.11/src/client.c
--- cvs-1.11.11.org/src/client.c	2004-01-18 12:28:32.162094080 +0100
+++ cvs-1.11.11/src/client.c	2004-01-18 12:30:35.439469099 +0100
@@ -4642,38 +4642,13 @@
     char *rsh_argv[10];
 
     if (!cvs_rsh)
-	/* People sometimes suggest or assume that this should default
-	   to "remsh" on systems like HPUX in which that is the
-	   system-supplied name for the rsh program.  However, that
-	   causes various problems (keep in mind that systems such as
-	   HPUX might have non-system-supplied versions of "rsh", like
-	   a Kerberized one, which one might want to use).  If we
-	   based the name on what is found in the PATH of the person
-	   who runs configure, that would make it harder to
-	   consistently produce the same result in the face of
-	   different people producing binary distributions.  If we
-	   based it on "remsh" always being the default for HPUX
-	   (e.g. based on uname), that might be slightly better but
-	   would require us to keep track of what the defaults are for
-	   each system type, and probably would cope poorly if the
-	   existence of remsh or rsh varies from OS version to OS
-	   version.  Therefore, it seems best to have the default
-	   remain "rsh", and tell HPUX users to specify remsh, for
-	   example in CVS_RSH or other such mechanisms to be devised,
-	   if that is what they want (the manual already tells them
-	   that).  */
-	cvs_rsh = "rsh";
+	cvs_rsh = "ssh";
     if (!cvs_server)
 	cvs_server = "cvs";
 
     /* The command line starts out with rsh. */
     rsh_argv[i++] = cvs_rsh;
 
-#   ifdef RSH_NEEDS_BINARY_FLAG
-    /* "-b" for binary, under OS/2. */
-    rsh_argv[i++] = "-b";
-#   endif /* RSH_NEEDS_BINARY_FLAG */
-
     /* Then we strcat more things on the end one by one. */
     if (root->username != NULL)
     {
@@ -4723,7 +4698,7 @@
     int child_pid;
 
     if (!cvs_rsh)
-	cvs_rsh = "rsh";
+	cvs_rsh = "ssh";
     if (!cvs_server)
 	cvs_server = "cvs";
 
