--- cvs-1.11.4/configure.in.orig	Fri Jan 17 19:46:05 2003
+++ cvs-1.11.4/configure.in	Fri Jan 17 19:55:47 2003
@@ -272,13 +272,12 @@
 AC_FUNC_MMAP
 
 # Try to find connect and gethostbyname.
-AC_CHECK_LIB(nsl, main)
 AC_SEARCH_LIBS(connect, xnet socket inet,
   AC_DEFINE(HAVE_CONNECT, 1,
 [Define if you have the connect function.]))
 dnl no need to search nsl for gethostbyname here since we should have
 dnl just added libnsl above if we found it.
-AC_SEARCH_LIBS(gethostbyname, netinet)
+AC_SEARCH_LIBS(gethostbyname, netinet nsl)
 
 
 dnl
