--- cvs-1.11.2/configure.in~	Wed May 15 21:33:03 2002
+++ cvs-1.11.2/configure.in	Wed May 15 21:35:54 2002
@@ -204,7 +204,7 @@
 
 AC_FUNC_FNMATCH
 if test "$ac_cv_func_fnmatch_works" = no; then
-  LIBOBJS="$LIBOBJS fnmatch.o"
+  AC_LIBOBJ(fnmatch)
 fi
 
 dnl for the buffer routine replacements
@@ -312,7 +312,7 @@
   AC_DEFINE(ENCRYPTION)
 fi
 
-AC_CHECK_FUNC(gethostname, :, LIBOBJS="$LIBOBJS hostname.o")
+AC_CHECK_FUNC(gethostname, :, AC_LIBOBJ(hostname))
 
 # Check for options requesting client and server feature. If none are
 # given and we have connect(), we want the full client & server arrangement.
@@ -366,7 +366,7 @@
 [AC_TRY_COMPILE([], [return __CYGWIN32__;],
 ccvs_cv_sys_cygwin32=yes, ccvs_cv_sys_cygwin32=no)])
 if test $ccvs_cv_sys_cygwin32 = yes; then
-  LIBOBJS="$LIBOBJS fncase.o"
+  AC_LIBOBJ(fncase)
   LIBS="$LIBS -ladvapi32"
 
   dnl On Windows you can only change file times if you can write to
