diff -ur cvs-1.11.1/configure.in cvs-1.11.1.zlib/configure.in
--- cvs-1.11.1/configure.in	Wed Apr 25 23:51:25 2001
+++ cvs-1.11.1.zlib/configure.in	Wed May 30 21:48:59 2001
@@ -300,6 +300,8 @@
   enable_server=yes
 fi])
 
+AC_CHECK_LIB(z, zlibVersion)
+
 dnl The auth server needs to be able to check passwords against passwd
 dnl file entries, so we only #define AUTH_SERVER_SUPPORT if we can
 dnl find the crypt function. 
@@ -410,8 +410,7 @@
 	  tools/Makefile \
 	  vms/Makefile \
 	  windows-NT/Makefile \
-	  windows-NT/SCC/Makefile \
-	  zlib/Makefile],
+	  windows-NT/SCC/Makefile]
 	  [chmod +x \
 	  contrib/clmerge \
 	  contrib/cln_hist \
diff -ur cvs-1.11.1/src/Makefile.am cvs-1.11.1.zlib/src/Makefile.am
--- cvs-1.11.1/src/Makefile.am	Tue Apr 24 20:14:53 2001
+++ cvs-1.11.1.zlib/src/Makefile.am	Wed May 30 21:47:19 2001
@@ -20,7 +20,7 @@
 # try and remove naming ocnflicts and fix Automake to allow particular includes
 # to be attached only to particular object files.  Short term fix is either or.
 ##INCLUDES = -I. -I.. -I$(srcdir) -I$(top_srcdir)/lib
-INCLUDES = -I$(top_srcdir)/lib -I$(top_srcdir)/diff -I$(top_srcdir)/zlib $(includeopt)
+INCLUDES = -I$(top_srcdir)/lib -I$(top_srcdir)/diff $(includeopt)
 
 bin_PROGRAMS = cvs
 bin_SCRIPTS = cvsbug
@@ -100,8 +100,7 @@
 
 cvs_LDADD = \
 	../diff/libdiff.a \
-	../lib/libcvs.a \
-	../zlib/libz.a
+	../lib/libcvs.a
 
 # extra clean targets
 # wish this could be distclean-hdr-local but it's not part of automake
--- cvs-1.11.2/Makefile.am~	Thu Sep 13 21:08:15 2001
+++ cvs-1.11.2/Makefile.am	Sat May 11 01:30:51 2002
@@ -17,7 +17,7 @@
 ## Subdirectories to run make in for the primary targets.
 # Unix source subdirs, where we'll want to run lint and etags:
 # This is a legacy variable from b4 Automake
-USOURCE_SUBDIRS = lib zlib diff src
+USOURCE_SUBDIRS = lib diff src
 # All other subdirs:
 SUBDIRS = $(USOURCE_SUBDIRS) man doc contrib tools \
 	windows-NT os2 emx vms
